In [4]:
from sklearn.datasets import load_iris

x, y = load_iris(return_X_y=True)

print("x shape:", x.shape)
print("y shape:", y.shape)

x shape: (150, 4)
y shape: (150,)


In [5]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y ,test_size= 0.2,random_state=42, stratify=y)

In [6]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier

In [7]:
models = {
    "Logistic Regression": LogisticRegression(max_iter=1000),
    "Decision Tree": DecisionTreeClassifier(random_state=42),
    "KNN": KNeighborsClassifier(),
    "Random Forest": RandomForestClassifier(
        n_estimators=100,
        random_state=42
    )
}

In [8]:
from sklearn.model_selection import cross_val_score

for name,model in models.items():

    scores = cross_val_score(
        model,
        x_train,
        y_train,
        cv= 5,
        scoring='accuracy'
    )

    print(
        f"{name:20} | "
        f"Mean CV Accuracy = {scores.mean():.4f}"
    )


Logistic Regression  | Mean CV Accuracy = 0.9667
Decision Tree        | Mean CV Accuracy = 0.9417
KNN                  | Mean CV Accuracy = 0.9750
Random Forest        | Mean CV Accuracy = 0.9500


In [9]:
for name, model in models.items():

    model.fit(x_train, y_train)

    accuracy = model.score(x_test, y_test)

    print(
        f"{name:20} | "
        f"Test Accuracy = {accuracy:.4f}"
    )

Logistic Regression  | Test Accuracy = 0.9667
Decision Tree        | Test Accuracy = 0.9333
KNN                  | Test Accuracy = 1.0000
Random Forest        | Test Accuracy = 0.9000


# 04. Model Comparison

## 1. Purpose of Model Comparison

Different machine learning algorithms can perform differently on the **same dataset**.

Instead of assuming one model is best, we can train multiple models and compare their performance using the **same evaluation procedure**.

```text
Same Dataset
     ↓
Multiple Models
     ↓
Same Evaluation Method
     ↓
Compare Performance
     ↓
Select Suitable Model
```

---

## 2. Models Compared

In our experiment, we compared:

```text
Logistic Regression
Decision Tree
KNN
Random Forest
```

---

## 3. Why Use Cross-Validation?

A single train/test split can give a result that depends on that particular split.

Cross-Validation evaluates each model across multiple validation splits.

We used:

```text
5-Fold Cross-Validation
```

Each model is evaluated using the same:

```text
x_train
y_train
cv=5
scoring="accuracy"
```

---

## 4. Model Comparison Code

```python
from sklearn.model_selection import cross_val_score

for name, model in models.items():

    scores = cross_val_score(
        model,
        x_train,
        y_train,
        cv=5,
        scoring="accuracy"
    )

    print(
        f"{name:20} | "
        f"Mean CV Accuracy = {scores.mean():.4f}"
    )
```

The process is:

```text
Model
  ↓
5-Fold Cross-Validation
  ↓
5 Validation Scores
  ↓
Calculate Mean
  ↓
Mean CV Accuracy
```

---

# 5. Our CV Results

```text
Logistic Regression  → 96.67%
Decision Tree        → 94.17%
KNN                  → 97.50%
Random Forest        → 95.00%
```

### Ranking

```text
1. KNN                 → 97.50% 🥇
2. Logistic Regression → 96.67%
3. Random Forest       → 95.00%
4. Decision Tree       → 94.17%
```

KNN had the highest mean CV accuracy among the models tested.

---

# 6. Final Test Set Comparison

After comparing models using Cross-Validation, we evaluated them on the **unseen test set**.

```python
for name, model in models.items():

    model.fit(x_train, y_train)

    accuracy = model.score(
        x_test,
        y_test
    )

    print(
        f"{name:20} | "
        f"Test Accuracy = {accuracy:.4f}"
    )
```

---

# 7. Our Test Results

```text
Logistic Regression  → 96.67%
Decision Tree        → 93.33%
KNN                  → 100.00%
Random Forest        → 90.00%
```

### Ranking

```text
1. KNN                 → 100.00% 🥇
2. Logistic Regression → 96.67%
3. Decision Tree       → 93.33%
4. Random Forest       → 90.00%
```

KNN correctly classified all **30 samples** in our test set.

```text
30 / 30 correct
     ↓
100% Test Accuracy
```

---

# 8. CV Accuracy vs Test Accuracy

Our results:

| Model               | Mean CV Accuracy | Test Accuracy |
| ------------------- | ---------------: | ------------: |
| Logistic Regression |           96.67% |        96.67% |
| Decision Tree       |           94.17% |        93.33% |
| KNN                 |       **97.50%** |   **100.00%** |
| Random Forest       |           95.00% |        90.00% |

The CV score is an **average across multiple validation folds**, while test accuracy comes from the particular unseen test set.

Therefore, the two scores don't have to be identical.

---

# 9. Model Comparison vs Hyperparameter Tuning

These two concepts are related but different.

### Hyperparameter Tuning

Find the best settings for **one model**.

```text
Decision Tree
     ↓
Try different max_depth
     ↓
Try different min_samples_split
     ↓
Cross-Validation
     ↓
Best Decision Tree configuration
```

### Model Comparison

Compare **different algorithms**.

```text
Logistic Regression
Decision Tree
KNN
Random Forest
       ↓
Cross-Validation
       ↓
Compare Performance
       ↓
Select Model
```

---

# 10. Important Point

Our result:

```text
KNN → 97.50% CV
KNN → 100% Test
```

does **not** mean:

> KNN is always the best machine learning algorithm.

It means:

> **For this Iris dataset, this particular train/test split, and these model configurations, KNN performed best in our experiment.**

For a real project, we would also consider things such as:

```text
Accuracy
Training time
Prediction time
Model complexity
Interpretability
Memory usage
Dataset size
```

So **highest accuracy is not always the only factor** when selecting a model.

---

# 11. Complete Model Selection Workflow

```text
Dataset
   ↓
Train / Test Split
   ↓
Choose Candidate Models
   ↓
Cross-Validation
   ↓
Compare CV Scores
   ↓
Select Promising Model
   ↓
Final Test Evaluation
   ↓
Choose Final Model
```

---

## ⭐ Key Points

```text
Model Comparison
→ Compare different algorithms using the same evaluation procedure.
```

```text
Cross-Validation
→ Helps compare models more reliably than relying on one split alone.
```

```text
CV Score
→ Used during model comparison/selection.
```

```text
Test Score
→ Final evaluation on unseen data.
```

```text
Model Comparison
→ Which algorithm performs better?
```

```text
Hyperparameter Tuning
→ Which settings are better for a model?
```

### One-line summary

> **Model comparison evaluates different algorithms under the same conditions so we can select a suitable model for the problem.**
